### PR TITLE
 MB-2591 Price Domestic Linehaul 

### PR DESCRIPTION
**Pull request by jacquelineIO**
## Description

Price linehaul service item.

The prime will request a payment request including service item Domestic Linehaul. The milmove system will receive the request, gather the params needed to price the service item. Create a payment request and price the payment request using the params previously gathered and return back the information (via a payment request) to the caller.

## Reviewer Notes

Similar PR was done to price Task Order Services. https://github.com/transcom/mymove/pull/4250 

## Setup

Testing information are in the steps below, but additional information to perform testing is in
* https://github.com/transcom/mymove/wiki/Acceptance-Testing-Payment-Requests  
* https://github.com/transcom/mymove/wiki/How-to-Test-the-Prime-API-(Local,-Staging,-and-Experimental)

## Formula for Dom Linehaul

Dom Linehaul Price Millicent  * distance * cwtWeight * escalationCompounded

## Verifying unit test calculations used in unit test if you are wanting to check the pricecents used for assertions. 

function `TestPriceDomesticLinehaul`
```
	dlhPriceCents      = unit.Cents(249770)
```
from file https://github.com/transcom/mymove/blob/d32e1dd618d06056ded679d061a9f02f2d91ad2b/pkg/services/ghcrateengine/domestic_linehaul_pricer_test.go#L17

```
//baseTotalPrice := weightPounds.ToCWTFloat64() * distanceMiles.Float64() * priceAndEscalation.PriceMillicents.Float64()
240000000 = 40.00 * 1200.00 * 5000.00

//escalatedTotalPrice := priceAndEscalation.EscalationCompounded * baseTotalPrice
249770400.00 = 1.040710 * 240000000.00 = 

// to cents
249770.4 = 249770400.00  / 1000
```

### Run 
```
make db_dev_reset db_dev_e2e_populate && rm bin/prime-api-client && make bin/prime-api-client
make server_run
```

### Using paylaod below create file `pr_create_dev.json`

```json
{                                                                                                                                                                            
  "body": {
    "isFinal": false,
    "moveTaskOrderID": "5d4b25bb-eb04-4c03-9a81-ee0398cb779e",
    "serviceItems": [
      {   
        "id": "9db1bf43-0964-44ff-8384-3297951f6781"
      }   
    ]   
  }
}
```

### Create a payment request. 

```sh
prime-api-client --insecure create-payment-request --filename pr_create_dev.json | jq
```

### View the results coming back from the payment request. 
Review the 9 params needed to price linehaul.

### Optionally, you can run `fetch-mto-updates` to pull the MTOs and view the same payment request info there. 

```sh
prime-api-client --insecure fetch-mto-updates | jq 'map(select(.id == "5d4b25bb-eb04-4c03-9a81-ee0398cb779e")) | .[0]'
```

### To verify the calculations, find the rate in your local dev database.

```sql
select price_millicents, escalation_compounded
from re_domestic_linehaul_prices dlp
         inner join re_contracts c on dlp.contract_id = c.id
         inner join re_contract_years cy on c.id = cy.contract_id
         inner join re_domestic_service_areas dsa on dlp.domestic_service_area_id = dsa.id
where c.code = 'TRUSS_TEST'
  and '2020-03-15 00:00:00 UTC' between cy.start_date and cy.end_date
  and dlp.is_peak_period = false
  and '1540' between dlp.weight_lower and dlp.weight_upper
  and '373' between dlp.miles_lower and dlp.miles_upper
  and dsa.service_area = '056';
```

### Verify the calculations from using e2e data with `prime-api-client` to create a payment-request. 

```
//baseTotalPrice := weightPounds.ToCWTFloat64() * distanceMiles.Float64() * priceAndEscalation.PriceMillicents.Float64()
1366545180 = (1540/100.0) * 373.0 * 237900

//escalatedTotalPrice := priceAndEscalation.EscalationCompounded * baseTotalPrice
1366545180.00 = 1.00 * 1366545180.00 = 

// to cents
1366545.18 = 1366545180.0 / 1000
```

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/wiki/backend#logging)
* [ ] The requirements listed in
 [Querying the Database Safely](https://github.com/transcom/mymove/wiki/backend#querying-the-database-safely)
 have been satisfied.
* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](https://github.com/transcom/mymove/wiki/migrate-the-database#zero-downtime-migrations))
  * [ ] Have been communicated to #g-database
  * [ ] Secure migrations have been tested following the instructions in our [docs](https://github.com/transcom/mymove/wiki/migrate-the-database#secure-migrations)
* [ ] There are no aXe warnings for UI.
* [ ] This works in [Supported Browsers and their phone views](https://github.com/transcom/mymove/tree/master/docs/adr/0016-Browser-Support.md) (Chrome, Firefox, IE, Edge).
* [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
* [ ] User facing changes have been reviewed by design.
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-2591) for this change

## Screenshots

If this PR makes visible UI changes, an image of the finished UI can help reviewers and casual
observers understand the context of the changes. A before image is optional and
can be included at the submitter's discretion.

Consider using an animated image to show an entire workflow instead of using multiple images. You may want to use GIPHY CAPTURE for this! 📸

_Please frame screenshots to show enough useful context but also highlight the affected regions._

### CODE REVIEW

File: pkg/services/ghcrateengine/domestic_linehaul_pricer.go [hunk 1]
```diff
@@ -1,71 +1,98 @@
...
-	if moveDate.IsZero() {
-		return 0, errors.New("MoveDate is required")
+	if requestedPickupDate.IsZero() {
+		return milliCentPriceAndEscalation{}, errors.New("MoveDate is required")
 	}
 	if distance <= 0 {
-		return 0, errors.New("Distance must be greater than 0")
+		return milliCentPriceAndEscalation{}, errors.New("Distance must be greater than 0")
 	}
 	if weight <= 0 {
```
Comment (jacquelineIO) [inline pkg/services/ghcrateengine/domestic_linehaul_pricer.go:88]: This should be the min weight of 500 to error check against

Comment (gidjin) [reply pkg/services/ghcrateengine/domestic_linehaul_pricer.go:88]: Should we be returning an error if the weight is less than the minimum? Or should we be just using 500 as the weight if it is less than 500 but not 0? Also where is the minimum being checked in the DLH pricer I'm not seeing it? I mostly want to know so I can handle it the same in DSH pricer.

File: pkg/services/ghcrateengine/param_convert.go [hunk 2]
```diff
@@ -13,6 +14,25 @@ const (
 	TimestampParamFormat = time.RFC3339
 )
 
+func getParamInt(params models.PaymentServiceItemParams, name models.ServiceItemParamName) (int, error) {
```
Comment (gidjin) [inline pkg/services/ghcrateengine/param_convert.go:17]: yay conflicts, I wrote one of these too. I'd ask for tests but I added tests to my branch already so it should all work out in the merge.

File: pkg/services/ghcrateengine/domestic_linehaul_pricer_test.go [hunk 3]
```diff
@@ -13,74 +14,64 @@ const (
...
-	domesticLinehaulPricer := NewDomesticLinehaulPricer(suite.DB(), suite.logger, testdatagen.DefaultContractCode)
-
-	suite.T().Run("success within peak period", func(t *testing.T) {
-		totalPrice, err := domesticLinehaulPricer.PriceDomesticLinehaul(
-			time.Date(testdatagen.TestYear, peakStart.month, peakStart.day, 0, 0, 0, 0, time.UTC),
-			dlhTestDistance, dlhTestWeight, dlhTestServiceArea)
+	paymentServiceItem := suite.setupDomesticLinehaulServiceItem()
+	linehaulServicePricer := NewDomesticLinehaulPricer(suite.DB())
 
+	suite.T().Run("success using PaymentServiceItemParams", func(t *testing.T) {
```
Comment (gidjin) [inline pkg/services/ghcrateengine/domestic_linehaul_pricer_test.go:27]: Should we have a test for below minimum weight but above 0?

Comment (jacquelineIO) [reply pkg/services/ghcrateengine/domestic_linehaul_pricer_test.go:27]: added a test for this

File: pkg/services/ghcrateengine/domestic_linehaul_pricer.go (hunk 1 above)
Comment (jacquelineIO) [reply pkg/services/ghcrateengine/domestic_linehaul_pricer.go]: I left a comment in your PR @gidjin, but I think we should be returning an error if less than minimum weight (500). Update my code to reflect this.

File: pkg/services/ghcrateengine/domestic_linehaul_pricer.go [hunk 4]
```diff
@@ -1,71 +1,103 @@
...
+func fetchDomesticLinehaulPrice(db *pop.Connection, contractCode string, requestedPickupDate time.Time, isPeakPeriod bool, distance int, weight int, serviceArea string) (milliCentPriceAndEscalation, error) {
+	// Validate parameters
+	if requestedPickupDate.IsZero() {
+		return milliCentPriceAndEscalation{}, errors.New("MoveDate is required")
+	}
+	if distance <= dlhPricerMinimumDistance {
+		return milliCentPriceAndEscalation{}, fmt.Errorf("distance must be greater than %d", dlhPricerMinimumDistance)
+	}
+	if weight <= dlhPricerMinimumWeight {
+		return milliCentPriceAndEscalation{}, fmt.Errorf("weight must be greater than %d", dlhPricerMinimumWeight)
```
Comment (gidjin) [inline pkg/services/ghcrateengine/domestic_linehaul_pricer.go:94]: Should this be `weight must be greater than or equal to %d`? I thought 500 was a valid weight.

Comment (jacquelineIO) [reply pkg/services/ghcrateengine/domestic_linehaul_pricer.go:94]: @gidjin  you are right!!! It was 0 before. If you are merging your PR do you think you can add this?

**Pull request merged**